plugins/GSdx/GSRendererHW.cpp

```diff
@@ -49,27 +49,31 @@ GSRendererHW::GSRendererHW(GSTextureCache* tc)
 
 void GSRendererHW::SetScaling()
 {
-	int buffer_size = max(m_context->FRAME.FBW * 64, m_regs->DISP[m_regs->PMODE.EN1 == 1 ? 0 : 1].DISPFB.FBW * 64);
+	int FrameBuffer_Width = max(m_context->FRAME.FBW * 64, m_regs->DISP[m_regs->PMODE.EN1 == 1 ? 0 : 1].DISPFB.FBW * 64);
 
-	//Only increase the buffer size, don't make it smaller, it breaks games (GH3)
+	// GS doesn't have a specific register for the FrameBuffer height. so we get the height
+	// from physical units of the display rectangle in case the game uses a heigher value of height.
+	int Assumed_FrameBuffer_height = max(512, GetDisplayRect().height());
+	int scale_factor = FrameBuffer_Width / GetDisplayRect().width();
 
-	// Also don't change the size for custom resolution (m_upscale_multiplier = 0).
-	if (m_upscale_multiplier && m_width < (buffer_size * m_upscale_multiplier)) {
-		m_tc->RemovePartial();
-	} else {
-		return;
-	}
+	// Assumption is that Framebuffer width uses some kind of scaling factor and rescales the size
+	// with respect to the display rectangle, do the same for assumed framebuffer height.
+	if (scale_factor >= 2)
+		Assumed_FrameBuffer_height *= scale_factor;
 
-	m_height = buffer_size < 1024 ? 512 : 1024;
+	bool good_RTSize = m_width == (FrameBuffer_Width * m_upscale_multiplier) && m_height == (Assumed_FrameBuffer_height * m_upscale_multiplier);
+	bool initialized_register_state = (m_context->FRAME.FBW * 64) > 1 && GetDisplayRect().height() > 1;
 
-	if (m_upscale_multiplier > 1) {
-		m_width = buffer_size * m_upscale_multiplier;
-		m_height *= m_upscale_multiplier;
-	}
-	
-	
-	if (m_upscale_multiplier)
-		printf("Frame buffer size set to  %dx%d (%dx%d)\n", (m_width / m_upscale_multiplier), (m_height / m_upscale_multiplier), m_width, m_height);
+	// No need to resize for native/custom resolutions as default size will be enough for native and we manually get RT Buffer size for custom.
+	// don't resize until the display rectangle and register states are stabilized.
+	if ( m_upscale_multiplier <= 1 || good_RTSize || !initialized_register_state)
+		return;
+	else
+		m_tc->RemovePartial();
```

Comment:
Since the if returns when true, you don't really need the else.

SIde note: good_RTSize, FrameBuffer_Width, initialized_register_state, Assumed_FrameBuffer_height - the variable naming seems inconsistent to me.